### PR TITLE
`benchmark` now pipes input into the closure

### LINKED ISSUE #7762: `benchmark` should pipe the input into the closure
### Describe the bug

`benchmark` currently can't be used to benchmark only part of a pipeline, because it ignores everything to the left of it.

### How to reproduce

Here's a real-world example:
```
〉curl -sL "https://www.gutenberg.org/files/11/11-0.txt" | benchmark { $in | str downcase | split words | uniq -c | sort-by count --reverse | first 10 } 
Error: nu::shell::only_supports_this_input_type (link)

  × Input type not supported.
   ╭─[entry #5:1:1]
 1 │ curl -sL "https://www.gutenberg.org/files/11/11-0.txt" | benchmark { $in | str downcase | split words | uniq -c | sort-by count --reverse | first 10 }
   ·                                                                      ─┬─   ──────┬─────
   ·                                                                       │          ╰── only string input data is supported
   ·                                                                       ╰── input type: nothing
```
As you can see, I obviously don't want to benchmark the `curl` call, but in order to structure this pipeline to do so, I have to break this into two commands.

### Expected behavior

See above. Also, generally speaking, I feel like anything taking a closure that does NOT pipe the input into it should really be piping something else (like iteration elements) instead of just `null`.

### Screenshots

_No response_

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.73.0                                   |
| branch             | main                                     |
| commit_hash        | 1291b647aee4a566a69fc5786b5437cff3976d1b |
| build_os           | windows-x86_64                           |
| build_target       | x86_64-pc-windows-msvc                   |
| rust_version       | rustc 1.65.0 (897e37553 2022-11-02)      |
| rust_channel       | 1.65.0-x86_64-pc-windows-msvc            |
| cargo_version      | cargo 1.65.0 (4bc8f24d3 2022-10-20)      |
| pkg_version        | 0.73.0                                   |
| build_time         | 2022-12-21 16:52:27 +10:00               |
| build_rust_channel | release                                  |
| features           | database, default, trash, which, zip     |
| installed_plugins  |                                          |

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by webbedspace**
# Description

Closes #7762. See issue for motivation.

# User-Facing Changes

Something like this is now possible without having to split this into 2 commands:
```
fetch "https://www.gutenberg.org/files/11/11-0.txt" | benchmark { str downcase | split words | uniq -c | sort-by count --reverse | first 10 }
```

# Tests + Formatting

Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass

# After Submitting

If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date.

### CODE REVIEW

File: crates/nu-command/src/system/benchmark.rs [hunk 1]
```diff
@@ -16,13 +16,17 @@ impl Command for Benchmark {
...
     fn signature(&self) -> nu_protocol::Signature {
         Signature::build("benchmark")
-            .required("block", SyntaxShape::Block, "the block to run")
-            .input_output_types(vec![(Type::Block, Type::String)])
+            .required(
+                "closure",
+                SyntaxShape::Closure(Some(vec![SyntaxShape::Any])),
+                "the closure to run",
+            )
+            .input_output_types(vec![(Type::Closure, Type::String)])
```
Comment (sholderbach) [inline crates/nu-command/src/system/benchmark.rs:29]: The input type in your examples is `Type::Any`/`Type::Nothing` depending on what you provide the closure in the argument block.
```suggestion
            .input_output_types(vec![(Type::Any, Type::Duration),(Type::Nothing, Type::Duration)])
```

File: crates/nu-command/src/system/benchmark.rs [hunk 2]
```diff
@@ -41,12 +45,27 @@ impl Command for Benchmark {
...
+        // and then converted back into a pipeline for eval_block().
+        // So, the metadata must be saved here and restored at that point.
+        let input_metadata = input.metadata();
+        let input_val = input.into_value(call.head);
+
+        if let Some(var) = block.signature.get_positional(0) {
+            if let Some(var_id) = &var.var_id {
+                stack.add_var(*var_id, input_val.clone());
+            }
+        }
```
Comment (sholderbach) [inline crates/nu-command/src/system/benchmark.rs:60]: Not sure how this affects benchmarking of commands that are streaming regularly?

But in general "caching" the input could also be beneficial for isolated microbenchmarks.

File: crates/nu-command/src/system/benchmark.rs [hunk 3]
```diff
@@ -41,12 +45,27 @@ impl Command for Benchmark {
...
+
+        // Get the start time after all other computation has been done.
         let start_time = Instant::now();
         eval_block(
             engine_state,
             &mut stack,
             block,
-            PipelineData::empty(),
+            input_val.into_pipeline_data_with_metadata(input_metadata),
```
Comment (sholderbach) [inline crates/nu-command/src/system/benchmark.rs:68]: Yeah that creates a `PipelineData::Value` and is not specialized for `Value::List` or `ListStream`

File: crates/nu-command/src/system/benchmark.rs [hunk 4]
```diff
@@ -41,12 +45,27 @@ impl Command for Benchmark {
         let redirect_stderr = call.redirect_stderr;
 
         let mut stack = stack.captures_to_stack(&capture_block.captures);
+
+        // In order to provide the pipeline as a positional, it must be converted into a value.
+        // But because pipelines do not have Clone, this one has to be cloned as a value
+        // and then converted back into a pipeline for eval_block().
+        // So, the metadata must be saved here and restored at that point.
```
Comment (sholderbach) [inline crates/nu-command/src/system/benchmark.rs:52]: Not 100% sure I can follow your comment here? The following two lines are concerned with the input and not the closure in the positional.

**Review by sholderbach — COMMENTED** [review]
Thanks that should be great to benchmark stuff independent from the test data generation!

File: crates/nu-command/src/system/benchmark.rs (hunk 4 above)
Comment (webbedspace) [reply crates/nu-command/src/system/benchmark.rs:52]: No, "the positional" means the positional inside the closure itself. `{|foo| ...}` ← `foo` is the positional. In order to provide the pipeline as a value for `foo` via `stack.add_var()`, it must be turned into `input_val`. `input_metadata` is also collected for reasons stated.

File: crates/nu-command/src/system/benchmark.rs (hunk 3 above)
Comment (webbedspace) [reply crates/nu-command/src/system/benchmark.rs:68]: This isn't supposed to work like an iterator command like `each`. For `[1 2 3] | benchmark { $in }`, `$in` is `[1 2 3]`, not `1`, `2` and `3` respectively. That means all stream inputs must be collected in order to be passed in as a value  (which is what `input.into_value()` does) and thus I believe producing `PipelineData::Value` is semantically correct.

**Comment by stormasm** [comment]
Great to see the benchmark work continue on improving...

@webbedspace good job on this one...

**Comment by webbedspace** [comment]
Is there anything else that needs to be done here?

**Comment by sholderbach** [comment]
> Is there anything else that needs to be done here?

I think we are good here. The comment initially confused the heck out of me but the code looks good!

**Pull request merged**